xtask/src/cli_extras.rs

```diff
@@ -0,0 +1,723 @@
+                            .collect();
+
+                        quote! {
+                            #enum_ident::#variant_ident(#(#field_bindings),*) => {
+                                match field_name {
+                                    "type" => Some(serde_json::Value::String(String::from(#variant_string))),
+                                    #(#field_matches,)*
+                                    _ => None,
+                                }
+                            }
+                        }
+                    }
+                }
+                Fields::Unit => {
+                    quote! {
+                        types::#enum_ident::#variant_ident => {
+                            match field_name {
+                                "type" => Some(serde_json::Value::String(String::from(#variant_string))),
+                                _ => None,
+                            }
+                        }
+                    }
+                }
+            }
+        })
+        .collect();
+
+    quote! {
+        type Item = Self;
+
+        fn field_names() -> &'static [&'static str] {
+            &[#(#field_strings),*]
+        }
+
+        fn get_field(&self, field_name: &str) -> Option<serde_json::Value> {
+            match self {
+                #(#variant_arms,)*
+            }
+        }
+
+        fn items(&self) -> Box<dyn Iterator<Item = &Self::Item> + '_> {
+            Box::new(std::iter::once(self))
+        }
+    }
+}
+
+/// Generate a `field_names` method for `CliCommand` use to determine which subcommands
+/// are eligible for the `--format` flag, and adding the list of available fields `--help`
+/// output.
+fn gen_cli_command(sdk_file: &syn::File, cli_file: &syn::File) -> TokenStream {
+    let builder_content = mod_content(sdk_file, "builder");
+
+    let cli_variants = cli_file
+        .items
+        .iter()
+        .filter_map(|i| match i {
+            Item::Enum(ItemEnum {
+                ident, variants, ..
+            }) if ident == "CliCommand" => Some(variants),
+            _ => None,
+        })
+        .next()
+        .unwrap();
+
+    // We expect tabular formatting to be useful when querying. Creation APIs in particular already
+    // have many arguments, adding the bulky `--format` help text on top of that will make them
+    // harder to understand. However, there is no 100% reliable way to determine which subcommand is
+    // performing a query. The return type for create and view endpoints are frequently the same, so
+    // we can't discriminate by type. Trying to limit to `List` and `View` will miss a fair number
+    // of query endpoints, e.g. `NetworkingBgpImportedRoutesIpv4` and `CurrentUserGroups`. The
+    // heuristic below should catch most new write operations.
+    let action_words = HashSet::from([
+        "Add", "Attach", "Create", "Demote", "Detach", "Probe", "Promote", "Reboot", "Resend",
+        "Start", "Stop", "Update",
```

Comment:
I go back and forth on whether this is a good idea. As the comment says, I think having `--format` for write actions will just add noise to the help test.

However, this list of words to filter on is inevitably going to fall out of date, so we'll end up with _most_ write operations not having the flag.